Adding more test cases

prashant-shahi opened the pull request:
Adding new test cases as per the internal discussion.

- [x] Type directive: s * * deletion
- [x] Type System: type function
- [x] Upsert: 1 mutation using Do
- [x] Upsert: 2 mutations using Do
- [x] Upsert: 1 mutation, 1 query using Do
- [x] Upsert: 1 query using Do
- [x] Upsert: 0 query, 0 mutation using Do
- [x] Upsert: Json Tests
- [x] Conditional Upsert
- [x] TxnConflict Exception
- [x] TxnReadOnly Exception
- [x] TxnFinished Exception
- [x] Bulk Set using Do
- [x] Bulk Delete using Do
- [x] ACL Test - Read
- [x] ACL Test - Write
- [x] ACL Test - Modify



<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph-js/70)
<!-- Reviewable:end -->

prashant-shahi committed:
Adding more tests

prashant-shahi committed:
Adding more test cases

- conditional upsert test
- bulk set and delete test
- conflict/exception test
- doRequest test

prashant-shahi committed:
Added test cases for type system/directive

prashant-shahi committed:
Updated upsert and typeSystem tests

prashant-shahi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/70)* status: 0 of 6 files reviewed, 2 unresolved discussions (waiting on @mangalaman93, @paulftw, and @pullrequest[bot])

___
*[tests/integration/doRequest.spec.ts, line 20 at r1](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQt-r1-20:-Low7IBQ9PpR9mXK-CRh:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/f7adede12b9bf83761903904c7805a81bcf08011/tests/integration/doRequest.spec.ts#L20)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

Nitpicking: The naming on these variables could be a bit more descriptive.
</blockquote></details>

Done.
___
*[tests/integration/upsert.spec.ts, line 302 at r1](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQu-r1-302:-Low7R3UEpymK73pJvt2:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/f7adede12b9bf83761903904c7805a81bcf08011/tests/integration/upsert.spec.ts#L302)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

I would give more context for this test, why exactly should it not upsert? This would give more explanation for users that edit this code in the future.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

prashant-shahi committed:
Adding ACL tests

prashant-shahi committed:
removing commented sleep function

prashant-shahi committed:
Merge remote-tracking branch 'origin' into prashant/adding-tests

paulftw reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/70)* status: 0 of 9 files reviewed, 16 unresolved discussions (waiting on @mangalaman93, @paulftw, @prashant-shahi, and @pullrequest[bot])

___
*[tests/integration/acl.spec.ts, line 7 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Ohgu0-aM0E7WTks3:-Lp8Ohgv8ymsujIlpHHr:bxzx2dj) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L7)):*
> ```XML
> import { createClient, createClientStub, setSchema, setup, wait } from "../helper";
> 
> const FIVE_MINUTES_IN_SECONDS = 5 * 60 * 1000; // 5 minutes in milliseconds
> ```

Five minutes is way too much for a simple test like this
___
*[tests/integration/acl.spec.ts, line 61 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8MzaCBsS2VQL5YNMZ:-Lp8MzaCBsS2VQL5YNM_:b-vv81xp) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L61)):*
> ```XML
>     const aclClientStub = createClientStub();
>     // tslint:disable-next-line no-console
>     console.log("Going to sleep");
> ```

no need for printing debug info to console
___
*[tests/integration/acl.spec.ts, line 62 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Mhgq4KfwAbA2E0_C:-Lp8Mhgq4KfwAbA2E0_D:bal4741) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L62)):*
> ```XML
>     // tslint:disable-next-line no-console
>     console.log("Going to sleep");
>     await wait(8000);
> ```

Having hardcoded timeouts is generally a bad idea, 8 seconds for a test like this is really long. We should either avoid a timeout or implement a periodic pulling for when the server is up.
___
*[tests/integration/acl.spec.ts, line 104 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8NCvH4faU91DGNmBA:-Lp8NCvH4faU91DGNmBB:b-n0gcwg) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L104)):*
> ```XML
>     await cmd(command);
>     // tslint:disable-next-line no-console
>     console.log("Going to sleep");
> ```

if waiting here is necessary we should file a bug against dgraph - once commandline tool has returned control the changes must've been completed.
___
*[tests/integration/acl.spec.ts, line 121 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8NSF-CrcNzJMRWbM3:-Lp8NSF-CrcNzJMRWbM4:b-uppunk) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L121)):*
> ```XML
>         const res: dgraph.Response = await txn.query(query);
>         // tslint:disable-next-line no-console
>         console.log(`res: ${JSON.stringify(res.getJson())}`);
> ```

console.log here and below
___
*[tests/integration/acl.spec.ts, line 162 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8NiUY5313vVDB2h9S:-Lp8NiUY5313vVDB2h9T:b-uy360u) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L162)):*
> ```XML
>         `);
>         await aclClient.alter(operation);
>         if (!expected) {
> ```

I think it'd be better to have this method return true/false, and then the assertion would become

await expect(tryAltering()).resolves.toBe(true)
await expect(tryAltering()).resolves.toBe(false)

currently the tryX methods all duplicate the same if(!expected) / if(expected) logic, on top of doing two separate things: altering the db and comparing actual vs expected result.
___
*[tests/integration/bulkSetDelete.spec.ts, line 13 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8RJjd8Mmg2mLbAbVv:-Lp8RJjd8Mmg2mLbAbVw:bws2zi9) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L13)):*
> ```XML
> };
> const profiles: Profile[] = [
>     { name: "Aman", email: "aman@dgraph.io", age: 26 },
> ```

I wouldn't use real emails of other people in tests, but if they are ok with that...
___
*[tests/integration/bulkSetDelete.spec.ts, line 40 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8P6pkE_7T_iW-csEP:-Lp8P6pkE_7T_iW-csEQ:b8b0bg6) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L40)):*
> ```XML
>     const mu = new dgraph.Mutation();
>     const mutationNquads = getMutationNquads();
>     mu.setSetNquads(mutationNquads);
> ```

how about mu.setSetNquads(getMutationNquads()) ?
___
*[tests/integration/bulkSetDelete.spec.ts, line 48 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8PKuL7kKY0P9rZq9w:-Lp8PKuL7kKY0P9rZq9x:b-nibtd5) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L48)):*
> ```XML
>     await txn.doRequest(req);
> 
>     const res = await checkIntegrity();
> ```

I thought checkIntegrity() would check integrity of the written data, but in reality it's just reading the stored data. needs a more descriptive name
___
*[tests/integration/bulkSetDelete.spec.ts, line 93 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Q73-E2u7wCVxWaSI:-Lp8Q73-E2u7wCVxWaSJ:bubrwfh) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L93)):*
> "perform bulk set using Do" 
what does "Do" refer to? is it doRequest or is it a "do" keyword in the new mutation syntax? In either case I think description can be more specific
___
*[tests/integration/bulkSetDelete.spec.ts, line 113 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8QUJY8KLGsNbGdAAm:-Lp8QUJY8KLGsNbGdAAn:b-z328n4) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L113)):*
> ```XML
> 
>         await doBulkSet();
>         await doBulkDelete();
> ```

It's not very clear from the code of this test where and how the results are checked. 
I'd do something like:

await doBulkSet()
await expect(readData()).resolves.toBe(goodData);

await doBulDelete()
await expect(readData()).resolves.toBe(noDataInDB);
___
*[tests/integration/conditionalUpsert.spec.ts, line 120 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8RobW7WynaAhK4lJT:-Lp8RobW7WynaAhK4lJU:bb5atk0) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/conditionalUpsert.spec.ts#L120)):*
> ```XML
> }
> 
> async function doNotConditionalUpsert(): Promise<void> {
> ```

How about doUnconditionalUpsert  / doUpsert / doInsert ?
___
*[tests/integration/typeSystem.spec.ts, line 82 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8To4qEC1sf9nEcARV:-Lp8To4qEC1sf9nEcARW:b4d6ilk) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/typeSystem.spec.ts#L82)):*
> ```XML
> 
>     it("should perform s * * delete", async () => {
>         const client = await setup();
> ```

I maybe didn't read this file carefully enough but neither perfromMutation nor performDeletion use declared type for mutations. Should this test go into another file?
___
*[tests/integration/typeSystem.spec.ts, line 94 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Su_f0rMDF3hltPBB:-Lp8Su_g88e_ZKcT6rrv:b-3vsup8) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/typeSystem.spec.ts#L94)):*
> ```XML
> 
>         await performMutation(client);
>         await performDeletion(client);
> ```

Same comment as bulk insert / delete.
I suggested it from the separation of concerns principle, in addition to that performMutation / performDeletion currently share quite a bit of code to construct a query, run it and check the result. this can be extracted into a shared function.


<!-- Sent from Reviewable.io -->

paulftw reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/70)* status: 0 of 9 files reviewed, 17 unresolved discussions (waiting on @mangalaman93, @paulftw, @prashant-shahi, and @pullrequest[bot])

*[a discussion](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8UXHy0CA0uU1L2Nr-:-Lp8UXHy0CA0uU1L2Nr0:b699y3g) (no related file):*
This PR is quite big, I suggest we break it down into multiple smaller PRs - one pull request per foo.spec.ts would be ideal

---


<!-- Sent from Reviewable.io -->

paulftw reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/70)* status: 0 of 9 files reviewed, 18 unresolved discussions (waiting on @mangalaman93, @paulftw, @prashant-shahi, and @pullrequest[bot])

___
*[tests/integration/acl.spec.ts, line 2 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8VAQ299q1nN0pEI4R:-Lp8VAQ299q1nN0pEI4S:bdn0aoq) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L2)):*
> ```XML
> import { exec } from "child_process";
> import * as util from "util";
> ```

import { promisify } from "util" and then you can simply write promisify(exec)


<!-- Sent from Reviewable.io -->

prashant-shahi committed:
Merge remote-tracking branch 'origin' into prashant/adding-tests

prashant-shahi committed:
Minor changes

- Reduced jest timeout to 1 minute per test
- Using dgraph address from helper file
- Removed debugging console logs

prashant-shahi committed:
Code cleanup

prashant-shahi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/70)* status: 0 of 9 files reviewed, 18 unresolved discussions (waiting on @mangalaman93, @paulftw, and @pullrequest[bot])

*[a discussion](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8UXHy0CA0uU1L2Nr-:-LpCwsgjBAZx9Em4v_HC:b-g2zwrg) (no related file):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

This PR is quite big, I suggest we break it down into multiple smaller PRs - one pull request per foo.spec.ts would be ideal
</blockquote></details>

There was single PR for the same number of tests in other clients, so I followed the same pattern.
Do let me know if I should break it down.

---
___
*[tests/integration/acl.spec.ts, line 2 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8VAQ299q1nN0pEI4R:-LpCxNBg3NJvM5FScOP3:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L2)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

import { promisify } from "util" and then you can simply write promisify(exec)
</blockquote></details>

Done.
___
*[tests/integration/acl.spec.ts, line 7 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Ohgu0-aM0E7WTks3:-LpCxTN6-RuqR41sV_jv:b1aok76) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L7)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

Five minutes is way too much for a simple test like this
</blockquote></details>

Changed it to 1 minute per test for `acl.spec.ts`.
___
*[tests/integration/acl.spec.ts, line 61 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8MzaCBsS2VQL5YNMZ:-LpCxo_0-qnB6I_x6O-8:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L61)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

no need for printing debug info to console
</blockquote></details>

Done.
___
*[tests/integration/acl.spec.ts, line 62 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Mhgq4KfwAbA2E0_C:-LpDDoc5CIracCbpAXfv:bapdwis) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L62)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

Having hardcoded timeouts is generally a bad idea, 8 seconds for a test like this is really long. We should either avoid a timeout or implement a periodic pulling for when the server is up.
</blockquote></details>

I have removed the delay here.

Right now, 6 seconds of delay is hardcoded after permission changes for ACL tests of other clients as well. We can implement periodic pulling is later in the future if need be.
___
*[tests/integration/acl.spec.ts, line 104 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8NCvH4faU91DGNmBA:-Lp9PLic0Gy0qNFRplpB:b-9f7w9j) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L104)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

if waiting here is necessary we should file a bug against dgraph - once commandline tool has returned control the changes must've been completed.
</blockquote></details>

Agreed. I had mentioned the issue to Manish.  He did mention to raise an issue in Dgraph.

I have raised an issue fo the same in Dgraph.
https://github.com/dgraph-io/dgraph/issues/4031
___
*[tests/integration/acl.spec.ts, line 121 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8NSF-CrcNzJMRWbM3:-LpCyGZK0VrLXttdGV4r:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L121)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

console.log here and below
</blockquote></details>

Done.
___
*[tests/integration/acl.spec.ts, line 162 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8NiUY5313vVDB2h9S:-LpDD_0D82FywYqVlcvB:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L162)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

I think it'd be better to have this method return true/false, and then the assertion would become

await expect(tryAltering()).resolves.toBe(true)
await expect(tryAltering()).resolves.toBe(false)

currently the tryX methods all duplicate the same if(!expected) / if(expected) logic, on top of doing two separate things: altering the db and comparing actual vs expected result.
</blockquote></details>

Done.
___
*[tests/integration/bulkSetDelete.spec.ts, line 13 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8RJjd8Mmg2mLbAbVv:-LpCzeri7m6Glwe3tu1Y:bk36g76) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L13)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

I wouldn't use real emails of other people in tests, but if they are ok with that...
</blockquote></details>

I did mention this to them. They were ok with it.
___
*[tests/integration/bulkSetDelete.spec.ts, line 40 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8P6pkE_7T_iW-csEP:-LpD-ID040gu_H-kUgrt:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L40)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

how about mu.setSetNquads(getMutationNquads()) ?
</blockquote></details>

Done.
___
*[tests/integration/bulkSetDelete.spec.ts, line 48 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8PKuL7kKY0P9rZq9w:-LpD-va2BgM-54w0zg6n:brrbt6n) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L48)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

I thought checkIntegrity() would check integrity of the written data, but in reality it's just reading the stored data. needs a more descriptive name
</blockquote></details>

Changed `checkIntegrity()` to `readProfiles()`.
___
*[tests/integration/bulkSetDelete.spec.ts, line 93 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Q73-E2u7wCVxWaSI:-LpD0UNz-o5iPRK5VZpH:b-nmtk8r) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L93)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

> "perform bulk set using Do" 
what does "Do" refer to? is it doRequest or is it a "do" keyword in the new mutation syntax? In either case I think description can be more specific
</blockquote></details>

I have changed the parent description to "Bulk operations using doRequest" and removed `using Do` from the tests' description.
___
*[tests/integration/bulkSetDelete.spec.ts, line 113 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8QUJY8KLGsNbGdAAm:-LpD0wHEC7eFBYErrEke:b-2a6a55) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L113)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

It's not very clear from the code of this test where and how the results are checked. 
I'd do something like:

await doBulkSet()
await expect(readData()).resolves.toBe(goodData);

await doBulDelete()
await expect(readData()).resolves.toBe(noDataInDB);
</blockquote></details>

`doBulkDelete()` deletes `name`, `email`, and `age` from `Prashant` node while other data remain intact.

There is an assertion at the end of the `doBulkSet()` and `doBulkDelete()` for the result checking. 
Do let me know if I need to modify something.
___
*[tests/integration/conditionalUpsert.spec.ts, line 120 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8RobW7WynaAhK4lJT:-LpD2Bnn0Hgm23ZOa2P6:b-eh8bfc) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/conditionalUpsert.spec.ts#L120)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

How about doUnconditionalUpsert  / doUpsert / doInsert ?
</blockquote></details>

Changed `doNotConditionalUpsert` to `doUnconditionalUpsert`.
___
*[tests/integration/typeSystem.spec.ts, line 82 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8To4qEC1sf9nEcARV:-LpD4vSw1cljB9cwWZ46:bfricg4) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/typeSystem.spec.ts#L82)):*

>const data: {
>        all: Person[];
>} = resp.getJson();

The declared type `Person` is used like this in *performDeletion* function. But maybe, this was unnecessary. Removing the type.
___
*[tests/integration/typeSystem.spec.ts, line 94 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Su_f0rMDF3hltPBB:-LpD9gWV5hydWRcTqZ7h:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/typeSystem.spec.ts#L94)):*
<details><summary><i>Previously, paulftw (Paul Korzhyk) wrote…</i></summary><blockquote>

Same comment as bulk insert / delete.
I suggested it from the separation of concerns principle, in addition to that performMutation / performDeletion currently share quite a bit of code to construct a query, run it and check the result. this can be extracted into a shared function.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

prashant-shahi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/70)* status: 0 of 9 files reviewed, 18 unresolved discussions (waiting on @mangalaman93, @paulftw, and @pullrequest[bot])

___
*[tests/integration/acl.spec.ts, line 7 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Ohgu0-aM0E7WTks3:-LpDGrLv-4XADIeThlb2:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L7)):*
<details><summary><i>Previously, prashant-shahi (Prashant Shahi) wrote…</i></summary><blockquote>

Changed it to 1 minute per test for `acl.spec.ts`.
</blockquote></details>

Done.
___
*[tests/integration/acl.spec.ts, line 62 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Mhgq4KfwAbA2E0_C:-LpDGsNpDLogCMeptCj-:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L62)):*
<details><summary><i>Previously, prashant-shahi (Prashant Shahi) wrote…</i></summary><blockquote>

I have removed the delay here.

Right now, 6 seconds of delay is hardcoded after permission changes for ACL tests of other clients as well. We can implement periodic pulling is later in the future if need be.
</blockquote></details>

Done.
___
*[tests/integration/acl.spec.ts, line 104 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8NCvH4faU91DGNmBA:-LpDGt9F-7LO7akt8XP7:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/acl.spec.ts#L104)):*
<details><summary><i>Previously, prashant-shahi (Prashant Shahi) wrote…</i></summary><blockquote>

Agreed. I had mentioned the issue to Manish.  He did mention to raise an issue in Dgraph.

I have raised an issue fo the same in Dgraph.
https://github.com/dgraph-io/dgraph/issues/4031
</blockquote></details>

Done.
___
*[tests/integration/bulkSetDelete.spec.ts, line 13 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8RJjd8Mmg2mLbAbVv:-LpDGu6JE30IPc5h0oq2:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L13)):*
<details><summary><i>Previously, prashant-shahi (Prashant Shahi) wrote…</i></summary><blockquote>

I did mention this to them. They were ok with it.
</blockquote></details>

Done.
___
*[tests/integration/bulkSetDelete.spec.ts, line 48 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8PKuL7kKY0P9rZq9w:-LpDGuvNErdZQoI_85w0:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L48)):*
<details><summary><i>Previously, prashant-shahi (Prashant Shahi) wrote…</i></summary><blockquote>

Changed `checkIntegrity()` to `readProfiles()`.
</blockquote></details>

Done.
___
*[tests/integration/bulkSetDelete.spec.ts, line 93 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8Q73-E2u7wCVxWaSI:-LpDGvKnCbSi9ieX22El:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L93)):*
<details><summary><i>Previously, prashant-shahi (Prashant Shahi) wrote…</i></summary><blockquote>

I have changed the parent description to "Bulk operations using doRequest" and removed `using Do` from the tests' description.
</blockquote></details>

Done.
___
*[tests/integration/bulkSetDelete.spec.ts, line 113 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8QUJY8KLGsNbGdAAm:-LpDGvtp8x2uv3unZXBs:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/bulkSetDelete.spec.ts#L113)):*
<details><summary><i>Previously, prashant-shahi (Prashant Shahi) wrote…</i></summary><blockquote>

`doBulkDelete()` deletes `name`, `email`, and `age` from `Prashant` node while other data remain intact.

There is an assertion at the end of the `doBulkSet()` and `doBulkDelete()` for the result checking. 
Do let me know if I need to modify something.
</blockquote></details>

Done.
___
*[tests/integration/conditionalUpsert.spec.ts, line 120 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8RobW7WynaAhK4lJT:-LpDGwQKAoauq8__EuvX:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/conditionalUpsert.spec.ts#L120)):*
<details><summary><i>Previously, prashant-shahi (Prashant Shahi) wrote…</i></summary><blockquote>

Changed `doNotConditionalUpsert` to `doUnconditionalUpsert`.
</blockquote></details>

Done.
___
*[tests/integration/typeSystem.spec.ts, line 82 at r2](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp8To4qEC1sf9nEcARV:-LpDGxFG1GETNd-Oy1k7:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/daeeba53e45cd0c6d5e15a2762c0b9e3b1f97d97/tests/integration/typeSystem.spec.ts#L82)):*
<details><summary><i>Previously, prashant-shahi (Prashant Shahi) wrote…</i></summary><blockquote>


>const data: {
>        all: Person[];
>} = resp.getJson();

The declared type `Person` is used like this in *performDeletion* function. But maybe, this was unnecessary. Removing the type.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

prashant-shahi commented on tests/integration/conflict.spec.ts:
```diff
@@ -24,4 +24,35 @@ describe("conflict", () => {
+        await expect(p1).resolves.toBeUndefined();
+
+        const p2 = txn1.commit();
+        await expect(p2).rejects.toBe(dgraph.ERR_FINISHED);
```
This test is expected to throw the `ERR_FINISHED` exception.

prashant-shahi committed:
Code cleanup and updated test labels

prashant-shahi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js/70)* status: 0 of 9 files reviewed, 47 unresolved discussions (waiting on @mangalaman93, @paulftw, @prashant-shahi, and @pullrequest[bot])

___
*[tests/integration/acl.spec.ts, line 106 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp3YGl5aegI9SbzjZv4-r3-106:-LpTTm8wB0IsLnROXW6n:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/acl.spec.ts#L106)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

Because this is catching any error, this could hide some types of problems. It could be that the test that is asserting that this will return `false` passes, but it could pass for the wrong reason. It would be a little safer if you ensure it's the write type of error. The same applies to the `tryWriting` and `tryAltering` functions below.
</blockquote></details>

Done.
___
*[tests/integration/acl.spec.ts, line 145 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp3YGl5aegI9SbzjZv4-r3-145:-LpTTtk4EGNUcy4bDZFE:b-ucfknh) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/acl.spec.ts#L145)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

These are nice looking tests. They're very easy to read and they tell a story about how these functions work.
</blockquote></details>

Thanks!
___
*[tests/integration/acl.spec.ts, line 146 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp3YGl5aegI9SbzjZv4-r3-146:-LpTUCFgCrRQ0lzyC8l7:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/acl.spec.ts#L146)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> only has read access
</blockquote></details>

Done.
___
*[tests/integration/acl.spec.ts, line 152 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Lp3YGl5aegI9SbzjZv4-r3-152:-LpTV-Ir3vCuEErheNZw:b-4gyl6g) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/acl.spec.ts#L152)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

If one of these tests fails, the last line to reset the permissions won't execute. This could cause subsequent tests to fail making it difficult to track down which test was the culprit. You could use an `afterEach` callback to ensure it's run regardless of if a test passes or fails.
</blockquote></details>

TBH, it is actually optional since we are setting the permissions before every test.
But since for the integrity, we are resetting the permissions at the end of each test, it makes sense to use `afterEach`.
___
*[tests/integration/bulkSetDelete.spec.ts, line 92 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe47-r3-92:-LpTWJVG5NpBjF6Sb7J-:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/bulkSetDelete.spec.ts#L92)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> successfully performs bulk set
</blockquote></details>

Done.
___
*[tests/integration/bulkSetDelete.spec.ts, line 103 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe47-r3-103:-LpTWJtUE7Vw3-_8kTgh:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/bulkSetDelete.spec.ts#L103)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> successfully performs bulk delete
</blockquote></details>

Done.
___
*[tests/integration/conditionalUpsert.spec.ts, line 172 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe48-r3-172:-LpTWXcb6ERj-9Cjvx_h:bn85s7s) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/conditionalUpsert.spec.ts#L172)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

Is there an expectation statement for this or is it TBD? If so please use `it.skip(...)`
</blockquote></details>

`checkUpsertIntegrity` function has the expectation statement for this.
___
*[tests/integration/conditionalUpsert.spec.ts, line 175 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe48-r3-175:-LpTX6lE2OPA4OVyw4h0:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/conditionalUpsert.spec.ts#L175)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> when @if condition is false, do not perform upsert
</blockquote></details>

Done.
___
*[tests/integration/conditionalUpsert.spec.ts, line 189 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe48-r3-189:-LpTX9KoB2_4aLOUtSB0:bn85s7s) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/conditionalUpsert.spec.ts#L189)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

Is there an expectation statement for this or is it TBD? If so please use `it.skip(...)`
</blockquote></details>

`checkUpsertIntegrity` function has the expectation statement for this.
___
*[tests/integration/conflict.spec.ts, line 6 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe46-r3-6:-LpTXS700ZTO7hqoooHP:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/conflict.spec.ts#L6)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> aborts when TxnConflict raises exception
</blockquote></details>

Done.
___
*[tests/integration/conflict.spec.ts, line 28 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe46-r3-28:-LpTXT8V5MVhhExMDc87:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/conflict.spec.ts#L28)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> aborts when TxnReadOnly raises exception
</blockquote></details>

Done.
___
*[tests/integration/conflict.spec.ts, line 38 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe46-r3-38:-LpTXYo-AD-4jbj-tFlT:bqkhqit) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/conflict.spec.ts#L38)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

Is there an expectation statement for this or is it TBD? If so please use `it.skip(...)`
</blockquote></details>

The `txn1.mutate(mu)` statement expected to return a promise rejected with `ERR_READ_ONLY` error.
___
*[tests/integration/conflict.spec.ts, line 41 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe46-r3-41:-LpTYuq7BmxZu26Ah6Dv:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/conflict.spec.ts#L41)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

abort when TxnFinished raises exception
</blockquote></details>

Done.
___
*[tests/integration/conflict.spec.ts, line 56 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe46-r3-56:-LpTYxAdCwjkumznoZ7t:ba3x0z9) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/conflict.spec.ts#L56)):*
<details><summary><i>Previously, prashant-shahi (Prashant Shahi) wrote…</i></summary><blockquote>

This test is expected to throw the `ERR_FINISHED` exception.
</blockquote></details>

The `txn1.commit()` statement is expected to return a promise rejected with `ERR_FINISHED` error.
___
*[tests/integration/doRequest.spec.ts, line 15 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQt-r3-15:-LpTZVXD2bHPSKXJd4sY:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/doRequest.spec.ts#L15)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> insert 3Quads(mutation) and then query finishes succesfully
</blockquote></details>

Done.
___
*[tests/integration/doRequest.spec.ts, line 49 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQt-r3-49:-LpT_KtE-gADZKokYq2t:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/doRequest.spec.ts#L49)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> fails with two mutations

Perhaps indicate why this should happen in the label?
</blockquote></details>

Done.
___
*[tests/integration/doRequest.spec.ts, line 69 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQt-r3-69:-LpT_Kda0-_6xWyCAv_F:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/doRequest.spec.ts#L69)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> fails with zero mutations

Perhaps indicate why this should happen in the label?
</blockquote></details>

Done.
___
*[tests/integration/typeSystem.spec.ts, line 53 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe49-r3-53:-LpTaC550SBTA0cZG-Kt:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/typeSystem.spec.ts#L53)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> query fields expand using type system
</blockquote></details>

Done.
___
*[tests/integration/typeSystem.spec.ts, line 65 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe49-r3-65:-LpT_mdt6yMKx3dUlwBW:b-nn3uhy) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/typeSystem.spec.ts#L65)):*
All the tests make use of `expandQuery` function which has the following expectation statement.

> expect(data.all.length).toEqual(personCount);
___
*[tests/integration/typeSystem.spec.ts, line 68 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe49-r3-68:-LpTaBfA7GIg-q5OcEBd:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/typeSystem.spec.ts#L68)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> performs s * * delete
</blockquote></details>

Done.
___
*[tests/integration/typeSystem.spec.ts, line 81 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-Louv7o0JoVsNzLAQe49-r3-81:-LpTaJhd5KQ8C-i-8r7R:b-nn3uhy) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/typeSystem.spec.ts#L81)):*
All the tests make use of `expandQuery` function which has the following expectation statement.

> expect(data.all.length).toEqual(personCount);
___
*[tests/integration/upsert.spec.ts, line 48 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQu-r3-48:-LpTbHw37dvoQxQoJxwh:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/upsert.spec.ts#L48)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

This is very similar to the `tryUpsert` function in `conditionalIUpsert.spec.ts`. Is there any way to make a shared helper function?
</blockquote></details>

Done.
___
*[tests/integration/upsert.spec.ts, line 331 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQu-r3-331:-LpTaWuM41YJYuBeb1rg:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/upsert.spec.ts#L331)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> update existing data and perform upsert
</blockquote></details>

Done.
___
*[tests/integration/upsert.spec.ts, line 341 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQu-r3-341:-LpTaXWsAkT0HVUMUqI7:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/upsert.spec.ts#L341)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> create new data and perform upsert
</blockquote></details>

Done.
___
*[tests/integration/upsert.spec.ts, line 351 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQu-r3-351:-LpTaYLh8dnRDM7Lj8k3:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/upsert.spec.ts#L351)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> successfully performs upsert loadset using Do
</blockquote></details>

Done.
___
*[tests/integration/upsert.spec.ts, line 369 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQu-r3-369:-LpTbYz5A0f_FCukfxoy:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/upsert.spec.ts#L369)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

> successfully perform upsert using JSON
</blockquote></details>

Done.
___
*[tests/integration/upsert.spec.ts, line 376 at r3](https://reviewable.io/reviews/dgraph-io/dgraph-js/70#-LoLfY6HKcvYAec0FtQu-r3-376:-LpTbter7C9cVvxT-qTm:b-2ny8o9) ([raw file](https://github.com/dgraph-io/dgraph-js/blob/a1d952e36b6774514d55d3e1f0fb9b0562f497ef/tests/integration/upsert.spec.ts#L376)):*
<details><summary><i>Previously, pullrequest[bot] wrote…</i></summary><blockquote>

Is there an expectation statement for this or is it TBD? If so please use `it.skip(...)`
</blockquote></details>

Here we have expectation statements inside `checkUpsertIntegrity` and `checkMutationIntegrity` functions, which `doJsonUpsert` function calls.


<!-- Sent from Reviewable.io -->

prashant-shahi committed:
Adding line break for long error messages

paulftw commented on tests/helper.ts:
```diff
@@ -43,3 +47,30 @@ export function wait(time: number): Promise<void> {
+    req.setMutationsList([mutation]);
+    req.setCommitNow(true);
+
+    let uid: string;
```
uid could have been declared below, during initialization

paulftw commented on tests/helper.ts:
```diff
@@ -43,3 +47,30 @@ export function wait(time: number): Promise<void> {
+    }
+}
+
+export const QUERY_PERMISSION_DENIED = new Error("7 PERMISSION_DENIED:\
```
These error constants shouldn't be in the shared helper file, as they are test specific (predicate names are hardcoded) and at least the first one is used only once

paulftw reviewed (approved):
Added a couple non-critical comments.
Approving as is, since I don't want to delay this PR further.
In the future please do break them up into multiple PRs as that'd be easier on everyone

prashant-shahi committed:
Moved errors from handler to acl test
